### PR TITLE
Remove A.tprod

### LINKED ISSUE #135: Substitute A.tprod by A'
I did it in a branch https://github.com/amontoison/Krylov.jl/tree/transpose_product
But we need to solve https://github.com/JuliaSmoothOptimizers/LinearOperators.jl/issues/92 before.

### PR DESCRIPTION
**Pull request by amontoison**
close #135 

I remove `A.tprod` but I don't understand something.
In the `REPL` with the following code :
```julia
using LinearOperators
M = rand(100,100)
x = rand(100)
A = PreallocatedLinearOperator(M)
A2 = A'
@allocated A2 * x
A3 = A'
@allocated A3 * x
```
Nothing is allocated when I compute `A3 * x` (cost of compilation of `*` method) whereas `Aᵀ * x` seems to be recompiled inside methods even if the method was compiled before.
And `Aᵀ = A'` is not allocating if the function has been "warm up"...

### CODE REVIEW

**Comment by dpo** [comment]
> whereas Aᵀ * x seems to be recompiled inside methods even if the method was compiled before.
And Aᵀ = A' is not allocating if the function has been "warm up"...

I don't understand. Could you give an example?

**Comment by amontoison** [comment]
This example shows that we don't allocate with the new transpose linear operator.
 
```julia
using LinearOperators

n  = 10^4
A  = rand(n, n)
x  = rand(n)
y  = rand(n)
op = PreallocatedLinearOperator(A) 

function test1(op, x, y, p=10)
	opᵀ = op'
	for i = 1:p
		p = op  * x
		q = opᵀ * y
	end
end

function test2(op, x, y, p=10)
	for i = 1:p
	  p = op * x
		q = op.tprod(y)
	end
end

test1(op, x, y)  # warm up 
@allocated test1(op, x, y)

test2(op, x, y)  # warm up
@allocated test2(op, x, y)

m = 100
A2 = rand(m, n)
op2 = PreallocatedLinearOperator(A2)
x2 = rand(n)
y2 = rand(m)

test1(op2, x2, y2)  # warm up 
@allocated test1(op2, x2, y2)

test2(op2, x2, y2)  # warm up
@allocated test2(op2, x2, y2)
```

**Comment by amontoison** [comment]
And this example shows that something else is allocated whereas I only changed `A.tprod` between `crmr1` and `crmr2`...

```julia
using LinearOperators, Printf, LinearAlgebra, SparseArrays

include("aux.jl")
include("crmr1.jl")
include("crmr2.jl")

L   = get_div_grad(32, 32, 32)
n   = size(L, 1)
m   = div(n, 2)
A   = PreallocatedLinearOperator(L)        # Dimension n x n
Au  = PreallocatedLinearOperator(L[1:m,:]) # Dimension m x n
Ao  = PreallocatedLinearOperator(L[:,1:m]) # Dimension n x m
b   = ones(n)                              # Dimension n
c   = ones(m)                              # Dimension m

crmr1(Au, c)  # warmup
@allocated crmr1(Au, c)

crmr2(Au, c)  # warmup
@allocated crmr2(Au, c)
```
[files.zip](https://github.com/JuliaSmoothOptimizers/Krylov.jl/files/3812689/files.zip)

**Comment by dpo** [comment]
According to the new TimedLinearOperators, there is no extra allocation:
```julia
julia> op = TimedLinearOperator(A);
julia> crmr1(op, b);
julia> op.timer
──────────────────────────────────────────────────────────────────
                           Time                   Allocations
                   ──────────────────────   ───────────────────────
 Tot / % measured:      7.02s / 3.72%            974KiB / 0.71%

 Section   ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────
 prod         441    139ms  53.1%   315μs     0.00B  0.00%    0.00B
 tprod        442    123ms  46.9%   277μs   6.91KiB  100%     16.0B
 ──────────────────────────────────────────────────────────────────
julia> op = TimedLinearOperator(A);
julia> crmr2(op, b);
julia> op.timer
──────────────────────────────────────────────────────────────────
                           Time                   Allocations
                   ──────────────────────   ───────────────────────
 Tot / % measured:      7.22s / 3.40%           16.5MiB / 0.04%

 Section   ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────
 prod         441    130ms  53.1%   295μs     0.00B  0.00%    0.00B
 ctprod       442    115ms  46.9%   260μs   6.91KiB  100%     16.0B
 ──────────────────────────────────────────────────────────────────
```
If I change `Aᵀ = A'` to `Aᵀ = transpose(A)` in `crmr2.jl`, I get the same:
```julia
julia> op = TimedLinearOperator(A);
julia> crmr2(op, b);
julia> op.timer
──────────────────────────────────────────────────────────────────
                           Time                   Allocations
                   ──────────────────────   ───────────────────────
 Tot / % measured:      88.8s / 0.29%           18.3MiB / 0.04%

 Section   ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────
 prod         441    135ms  52.8%   307μs     0.00B  0.00%    0.00B
 tprod        442    121ms  47.2%   273μs   6.91KiB  100%     16.0B
 ──────────────────────────────────────────────────────────────────
```
I do observe the same extra allocations as you do with `@allocated` and also with `@btime`. I'm not sure where they come from.

**Comment by amontoison** [comment]
I rebased my pull request. We should be able to use `Krylov.jl` with the last release of `LinearOperators.jl` (v0.7.1) with this commit.

File: Project.toml [hunk 1]
```diff
@@ -11,5 +11,5 @@ SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [compat]
-LinearOperators = "0.5.2, 0.6"
+LinearOperators = "≥ 0.7.1"
```
Comment (abelsiqueira) [inline Project.toml]: Use `LinearOperators = "0.7.1, 1"`, so it has an upper limit also.

**Comment by dpo** [comment]
🤞

**Comment by amontoison** [comment]
Wow, Travis / Appveyor tests time has been reduced by a factor 4  with our recent updates of `LinearOperators` and this commit !

**Comment by dpo** [comment]
The type annotations must be doing something right.

**Comment by dpo** [comment]
🤘

**Pull request merged**